clients/pkg/promtail/targets/file/decompresser.go

```diff
@@ -0,0 +1,298 @@
+	go decompressor.updatePosition()
+	metrics.filesActive.Add(1.)
+	return decompressor, nil
+}
+
+// mountReader instantiate a reader ready to be used by the decompressor.
+//
+// The selected reader implementation is based on the extension of the given file name.
+// It'll error if the extension isn't supported.
+func mountReader(f *os.File, logger log.Logger) (reader io.Reader, err error) {
+	ext := filepath.Ext(f.Name())
+	var decompressLib string
+
+	if strings.Contains(ext, "gz") { // .gz, .tar.gz
+		decompressLib = "compress/gzip"
+		reader, err = gzip.NewReader(f)
+	} else if ext == ".z" {
+		decompressLib = "compress/zlib"
+		reader, err = zlib.NewReader(f)
+	} else if ext == ".bz2" {
+		decompressLib = "bzip2"
+		reader = bzip2.NewReader(f)
+	}
+
+	level.Debug(logger).Log("msg", fmt.Sprintf("using %q to decompress file %q", decompressLib, f.Name()))
+
+	if reader != nil {
+		return reader, nil
+	}
+
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
+
+	supportedExtsList := strings.Builder{}
+	for ext := range supportedCompressedFormats() {
+		supportedExtsList.WriteString(ext)
+	}
+	return nil, fmt.Errorf("file %q has unsupported extension, it has to be one of %q", f.Name(), supportedExtsList.String())
+}
+
+func (t *decompressor) updatePosition() {
+	positionSyncPeriod := t.positions.SyncPeriod()
+	positionWait := time.NewTicker(positionSyncPeriod)
+	defer func() {
+		positionWait.Stop()
+		level.Info(t.logger).Log("msg", "position timer: exited", "path", t.path)
+		close(t.posdone)
+	}()
+
+	for {
+		select {
+		case <-positionWait.C:
+			if err := t.MarkPositionAndSize(); err != nil {
+				level.Error(t.logger).Log("msg", "position timer: error getting position and/or size, stopping decompressor", "path", t.path, "error", err)
+				return
+			}
+		case <-t.posquit:
+			return
+		}
+	}
+}
+
+// readLines read all existing lines of the given compressed file.
+//
+// It first decompress the file as a whole using a reader and then it will iterate
+// over its chunks, separated by '\n'.
+// During each iteration, the parsed and decoded log line is then sent to the API with the current timestamp.
+func (t *decompressor) readLines() {
+	level.Info(t.logger).Log("msg", "read lines routine: started", "path", t.path)
+	t.running.Store(true)
+
+	defer func() {
+		t.cleanupMetrics()
+		level.Info(t.logger).Log("msg", "read lines routine finished", "path", t.path)
+		close(t.done)
+	}()
+	entries := t.handler.Chan()
+
+	f, err := os.Open(t.path)
+	if err != nil {
+		level.Error(t.logger).Log("msg", "error reading file", "path", t.path, "error", err)
+		return
+	}
+	defer f.Close()
+
+	r, err := mountReader(f, t.logger)
+	if err != nil {
+		level.Error(t.logger).Log("msg", "error mounting new reader", "err", err)
+		return
+	}
+
+	level.Info(t.logger).Log("msg", "successfully mounted reader", "path", t.path, "ext", filepath.Ext(t.path))
+
+	megabyte := 1000000
+	maxLoglineSize := 2 * megabyte
+	buffer := make([]byte, 3*megabyte)
```

Comment:
Comments to explain rationale please

Reply:
I think a 4096 byte buffer would be more performant here; let's start with that, and in a subsequent PR we can try improve performance or make this configurable.